src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -418,6 +543,132 @@ MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToDisk(
     return new_data_part;
 }
 
+MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToS3(
+    const String & part_name,
+    const String & replica_path,
+    bool to_detached,
+    const String & tmp_prefix_,
+    const Disks & disks_s3,
+    PooledReadWriteBufferFromHTTP & in
+    )
+{
+    if (disks_s3.empty())
+        throw Exception("No S3 disks anymore", ErrorCodes::LOGICAL_ERROR);
+
+    auto disk = disks_s3[0];
+
+    static const String TMP_PREFIX = "tmp_fetch_";
+    String tmp_prefix = tmp_prefix_.empty() ? TMP_PREFIX : tmp_prefix_;
+
+    String part_relative_path = String(to_detached ? "detached/" : "") + tmp_prefix + part_name;
+    String part_download_path = data.getRelativeDataPath() + part_relative_path + "/";
+
+    if (disk->exists(part_download_path))
+        throw Exception("Directory " + fullPath(disk, part_download_path) + " already exists.", ErrorCodes::DIRECTORY_ALREADY_EXISTS);
+
+    CurrentMetrics::Increment metric_increment{CurrentMetrics::ReplicatedFetch};
+
+    disk->createDirectories(part_download_path);
+
+    size_t files;
+    readBinary(files, in);
+
+    auto volume = std::make_shared<SingleDiskVolume>("volume_" + part_name, disk);
+    MergeTreeData::MutableDataPartPtr new_data_part = data.createPart(part_name, volume, part_relative_path);
+
+    for (size_t i = 0; i < files; ++i)
+    {
+        String file_name;
+        UInt64 file_size;
+
+        readStringBinary(file_name, in);
+        readBinary(file_size, in);
+
+        String data_path = new_data_part->getFullRelativePath() + file_name;
+        String metadata_file = fullPath(disk, data_path);
+
+        {
+            auto file_out = createWriteBufferFromFileBase(metadata_file, 0, 0, DBMS_DEFAULT_BUFFER_SIZE, -1);
+
+            HashingWriteBuffer hashing_out(*file_out);
+
+            copyData(in, hashing_out, file_size, blocker.getCounter());
+
+            if (blocker.isCancelled())
+            {
+                /// NOTE The is_cancelled flag also makes sense to check every time you read over the network,
+                /// performing a poll with a not very large timeout.
+                /// And now we check it only between read chunks (in the `copyData` function).
+                disk->removeSharedRecursive(part_download_path, true);
+                throw Exception("Fetching of part was cancelled", ErrorCodes::ABORTED);
+            }
+
+            MergeTreeDataPartChecksum::uint128 expected_hash;
+            readPODBinary(expected_hash, in);
+
+            if (expected_hash != hashing_out.getHash())
+            {
+                throw Exception("Checksum mismatch for file " + metadata_file + " transferred from " + replica_path,
+                    ErrorCodes::CHECKSUM_DOESNT_MATCH);
+            }
+        }
+
+        if (!i)
+        { /// Check access for first s3 object of first file
+            if (!disk->checkFile(data_path))
```

Comment:
Next block is a bit of confusion. I'll try to rewrite it with direct sending s3 object name to check access.